File: Python/백준/Silver/11047. 동전 0/동전 0.py
import sys
N, K = map(int, sys.stdin.readline().split())
inp = []
for _ in range(N):
	tmp = int(sys.stdin.readline())
	if tmp <= K:
		inp.append(tmp)
cnt = 0
for i in range(len(inp) - 1, -1, -1):
	if K >= inp[i]:
		tmp = K // inp[i]
		K -= inp[i] * tmp
		cnt += tmp
print(cnt)

File: Java/백준/Silver/11047. 동전 0/동전 0.java
import java.io.*;
import java.util.*;

class Solution {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public void solution() throws IOException{
        String inp[] = br.readLine().split(" ");
        int N = Integer.parseInt(inp[0]);
        int total = Integer.parseInt(inp[1]);

        int coinsCnt = 0;
        int coins[] = new int[N];
        for (int i = 0; i < N; i++) {
            int coin = Integer.parseInt(br.readLine());
            coins[i] = coin;
            if (coins[i] <= total) coinsCnt += 1;
        }

        int ans = 0;
        for (int i = coinsCnt - 1; i > -1; i--) {
            if (total < 0) break;
            int divCnt = total / coins[i];
            if (divCnt > 0) {
                ans += divCnt;
                total -= coins[i] * divCnt;
            }
        }
        bw.write(ans + "\n");
        bw.flush();
        bw.close();
    }
}

public class Main {
    public static void main(String argv[]) throws IOException{
        Solution sol = new Solution();
        sol.solution();
    }
}